MapleServer2/PacketHandlers/Game/TradeHandler.cs

```diff
@@ -0,0 +1,261 @@
+        if (player is null)
+        {
+            return;
+        }
+
+        session.Send(TradePacket.RequestSent());
+        player.Session?.Send(TradePacket.TradeRequest(session.Player.Name, session.Player.CharacterId));
+    }
+
+    private static void HandleRequestRespond(PacketReader packet)
+    {
+        long otherCharacterId = packet.ReadLong();
+        // no idea what this does
+    }
+
+    private static void HandleAcceptRequest(GameSession session, PacketReader packet)
+    {
+        long otherCharacterId = packet.ReadLong();
+        Player otherPlayer = GameServer.PlayerManager.GetPlayerById(otherCharacterId);
+        if (otherPlayer is null)
+        {
+            return;
+        }
+
+        // Create Trade inventory
+        session.Player.TradeInventory = new(otherPlayer);
+        otherPlayer.TradeInventory = new(session.Player);
+
+        session.Send(TradePacket.RequestAccepted(otherCharacterId));
+        otherPlayer.Session?.Send(TradePacket.RequestAccepted(session.Player.CharacterId));
+    }
+
+    private static void HandleDeclineRequest(GameSession session, PacketReader packet)
+    {
+        long otherCharacterId = packet.ReadLong();
+        Player player = GameServer.PlayerManager.GetPlayerById(otherCharacterId);
+        player?.Session?.Send(TradePacket.RequestDeclined(session.Player.Name));
+    }
+
+    private static void HandleCloseTrade(GameSession session)
+    {
+        TradeInventory tradeInventory = session.Player.TradeInventory;
+        if (tradeInventory is null)
+        {
+            return;
+        }
+
+        Player otherPlayer = tradeInventory.OtherPlayer;
+
+        // Return items back to player's inventory
+        session.Player.TradeInventory.SendItems(session.Player, false);
+        session.Player.TradeInventory = null;
+
+        if (otherPlayer?.TradeInventory is null)
+        {
+            return;
+        }
+
+        session.Send(TradePacket.TradeStatus(false));
+        otherPlayer.Session?.Send(TradePacket.TradeStatus(false));
+    }
+
+    private static void HandleAddItemToTrade(GameSession session, PacketReader packet)
+    {
+        long itemUid = packet.ReadLong();
+        int amount = packet.ReadInt();
+
+        if (!session.Player.Inventory.HasItem(itemUid) || session.Player.TradeInventory.IsLocked)
+        {
+            return;
+        }
+
+        Item item = session.Player.Inventory.GetByUid(itemUid);
+
+        if (item.IsBound())
+        {
+            return;
+        }
+
+        // Split item
+        if (item.Amount > amount)
+        {
+            item.TrySplit(amount, out Item splitItem);
+            session.Send(ItemInventoryPacket.UpdateAmount(itemUid, item.Amount));
+            item = splitItem;
+        }
+        else
+        {
+            session.Player.Inventory.RemoveItem(session, itemUid, out Item removedItem);
+            item = removedItem;
+        }
```

Comment:
Can't you use **Inventory.Consume** here?

Reply:
No because we're not deleting the item. I also need to extract the item out of said stack.